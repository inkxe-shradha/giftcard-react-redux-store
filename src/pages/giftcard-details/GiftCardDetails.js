import {
  Button,
  Container,
  Divider,
  Grid,
  Paper,
  TextField,
  Typography,
} from "@mui/material";
import { makeStyles } from "@mui/styles";
import { Box } from "@mui/system";
import React, { useEffect, useRef, useState } from "react";
import { useParams } from "react-router-dom";
import "./index.scss";
import { useDispatch, useSelector } from "react-redux";
import {
  addNewComment,
  loadSingleCard,
  sendEmail,
} from "../../store/actions/giftCardAction";
import moment from "moment";
import GiftComment from "../../components/GiftComment/GiftComment";
import StarRatingComponent from "react-star-rating-component";
import SendEmailModal from "../../components/SendEmail/SendEmailModal";
import { toast } from "react-toastify";

const styles = makeStyles((theme) => ({
  paper: {
    padding: theme.spacing(2),
    border: "1px solid" + theme.palette.primary.main,
    wordBreak: "break-all",
    borderRadius: "10px",
    width: "100%",
  },
  maxImageHeight: {
    maxHeight: "500px",
    overflow: "auto",
  },
}));
const GiftCardDetails = (props) => {
  const classes = styles();
  const giftCardDetails = useSelector(
    (state) => state.giftCard.giftCardDetails
  );

  const { emailLoader, emailError } = useSelector((state) => ({
    emailLoader: state.giftCard.sendingEmailLoader,
    emailError: state.giftCard.giftCardError,
  }));

  const authenticationStatus = useSelector(
    (state) => state.auth.isAuthenticated
  );

  const userObject = useSelector((state) => state.auth.user);
  const dispatch = useDispatch();
  const [starCount, setStarCount] = useState(0);
  const [comment, setCommentCount] = useState("");
  const [isCommentEmpty, setCommentEmpty] = useState(false);
  const [isSendModalOpen, setIsSendModalOpen] = useState(false);
  const [errorMess, setErrorMess] = useState(emailError || "");
  const [editCommentId, setEditCommentId] = useState(null);
  const commentRef = useRef(null);

  const { id } = useParams();
  useEffect(() => {
    dispatch(loadSingleCard(id));
  }, [id, dispatch]);

  const handelAddComment = (e) => {
    if (!comment) {
      setCommentEmpty(true);
      return;
    }
    let commentObj = {};
    if (!editCommentId) {
      commentObj = {
        name: userObject.name,
        email: userObject.email,
        rating: starCount,
        comment,
        commented_on: moment().format("YYYY-MM-DD"),
        id: new Date().getTime() + Math.floor(Math.random() * 1000 + 1000),
      };
    }
    setCommentEmpty(false);
    dispatch(
      addNewComment({
        ...giftCardDetails,
        cardComments: !editCommentId
          ? [...giftCardDetails.cardComments, commentObj]
          : giftCardDetails.cardComments.map((ele) => {
              if (ele.id === editCommentId) {
                return { ...ele, comment, rating: starCount };
              }
              return ele;
            }),
      })
    ).then(() => {
      setCommentCount("");
      setStarCount(0);
      setEditCommentId(null);
    });
  };

  const handelSendEmail = (value) => {
    if (
      +userObject.balance_points < +giftCardDetails.cardPoints ||
      +giftCardDetails.cardCount === 0
    ) {
      setErrorMess("Insufficient balance points");
      return;
    }
    dispatch(
      sendEmail(
        value.email,
        userObject.email,
        `Your Gift Card Worth ${giftCardDetails.cardPoints} is here.`,
        giftCardDetails
      )
    )
      .then((res) => {
        console.log(res);
        setErrorMess("");
        setIsSendModalOpen(false);
        toast.success("Email sent successfully");
      })
      .catch((err) => {
        console.log(err);
        setErrorMess("Email not sent");
      });
  };

  const onRemoveComments = (id) => {
    const newComments = giftCardDetails.cardComments.filter(
      (comment) => comment.id !== id
    );
    dispatch(
      addNewComment({
        ...giftCardDetails,
        cardComments: newComments,
      })
    );
  };

  const onEditComments = (comment) => {
    setCommentCount(comment.comment);
    setStarCount(comment.rating);
    setEditCommentId(comment.id);
    commentRef.current.focus();
    commentRef.current.scrollIntoView({ behavior: "smooth" });
  };
  return (
    <Container maxWidth="xl" className="mt-5">
      {giftCardDetails && (
        <Paper className={classes.paper}>
          <Grid container spacing={4}>
            <Grid item xs={12} md={4} sm={6}>
              <Grid container spacing={2}>
                <Grid item xs={12} sm={12}>
                  <img
                    src={giftCardDetails.cardImage}
                    alt="Card Img Details"
                    className="img-fluid img-thumbnail"
                    width="100%"
                  />
                </Grid>
                <Grid item xs={12} sm={12}>
                  <Divider className="mb-2" />
                  <Button
                    fullWidth
                    onClick={() => setIsSendModalOpen(true)}
                    variant="contained"
                    className={!authenticationStatus ? "d-none" : ""}
                  >
                    Send The Gift
                  </Button>
                  <table className="table table-dark mt-1 table-borderless">
                    <tbody>
                      <tr>
                        <td>
                          <Typography variant="p" color="primary">
                            <strong>Gift Card Number</strong>
                          </Typography>
                        </td>
                        <td>#GIFT-{(giftCardDetails?.id || 101) + 2022}</td>
                      </tr>
                      <tr>
                        <td>
                          <Typography variant="p" color="primary">
                            <strong>Gift Card Value</strong>
                          </Typography>
                        </td>
                        <td>$ {giftCardDetails.cardPoints}</td>
                      </tr>
                      <tr>
                        <td>
                          <Typography variant="p" color="primary">
                            <strong>Gift Card Expiry Date</strong>
                          </Typography>
                        </td>
                        <td>
                          {moment(giftCardDetails.cardExpiryDate).format(
                            "MMMM Do YYYY"
                          )}
                        </td>
                      </tr>
                      <tr>
                        <td>
                          <Typography variant="p" color="primary">
                            <strong>Available Cards</strong>
                          </Typography>
                        </td>
                        <td>
                          {giftCardDetails.cardCount > 20 ? (
                            <span className="badge bg-gradient badge-pill bg-success">
                              {giftCardDetails.cardCount}
                            </span>
                          ) : (
                            <span className="badge bg-gradient badge-pill bg-danger">
                              {giftCardDetails.cardCount}
                            </span>
                          )}
                        </td>
                      </tr>
                    </tbody>
                  </table>
                </Grid>
              </Grid>
            </Grid>
            <Grid item xs={12} md={8} sm={6}>
              <Box>
                <Typography variant="h4">
                  {giftCardDetails.cardName} -{" "}
                  <small> {giftCardDetails.cardRetailer}</small>
                </Typography>
              </Box>
              {/* Product Description */}
              <Box className="mt-2">
                <Typography variant="body">
                  {giftCardDetails.cardLongDesc}
                </Typography>
              </Box>
              <Divider className="my-2" />
              {/* Comment Section */}
              <Box className={classes.maxImageHeight}>
                <Typography variant="h6">Rating & Reviews</Typography>
                {giftCardDetails?.cardComments?.length > 0 &&
                  giftCardDetails.cardComments.map((comment) => (
                    <GiftComment
                      comment={comment}
                      key={comment.id}
                      currentUser={userObject}
                      onRemove={onRemoveComments}
                      onEdit={onEditComments}
                    />
                  ))}
                {/* No comments */}
                {giftCardDetails?.cardComments?.length === 0 && (
                  <Typography variant="body" margin="normal" className="mt-2">
                    <em> No comments yet. Be the first to comment.</em>
                  </Typography>
                )}
              </Box>
              {authenticationStatus && (
                <Paper
                  elevation={10}
                  sx={{
                    padding: (theme) => {
                      return theme.spacing(2);
                    },
                    marginTop: (theme) => {
                      return theme.spacing(2);
                    },
                  }}
                >
                  <div className="my-2">
                    <Typography variant="h6">Add Comment</Typography>
                    <Typography variant="body">
                      Please add your comment here.
                    </Typography>
                  </div>
                  <StarRatingComponent
                    name="rating"
                    starCount={5}
                    value={starCount}
                    emptyStarColor="#ffff"
                    editing={true}
                    onStarClick={(nextValue) => setStarCount(nextValue)}
                  />
                  <TextField
                    inputRef={commentRef}
                    id="outlined-pass-comments-static"
                    label="Your valuable comments..."
                    aria-label="your comments"
                    fullWidth
                    multiline
                    rows={4}
                    name="comment"
                    required
                    value={comment}
                    error={isCommentEmpty}
                    onChange={(e) => setCommentCount(e.target.value)}
                  />
                  <Button
                    fullWidth
                    variant="contained"
                    onClick={handelAddComment}
                    className={
                      !authenticationStatus ? "d-none" : "btn-comment mt-2"
                    }
                  >
                    {editCommentId ? "Edit Comment" : "Add Comment"}
                  </Button>
                </Paper>
              )}
            </Grid>
          </Grid>
        </Paper>
      )}
      <SendEmailModal
        isOpen={isSendModalOpen}
        handleClose={() => setIsSendModalOpen(false)}
        handleSubmit={handelSendEmail}
        loading={emailLoader}
        errorMessage={errorMess}
      />
    </Container>
  );
};

export default GiftCardDetails;
