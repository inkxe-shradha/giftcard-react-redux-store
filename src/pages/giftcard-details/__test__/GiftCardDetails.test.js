import { createMemoryHistory } from "history";
import { mount } from "enzyme";
import { ThemeProvider } from "@mui/system";
import { BrowserRouter as Router } from "react-router-dom";
import { createTheme } from "@mui/material";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import GiftCardDetails from "../GiftCardDetails";
import * as reactRedux from "react-redux";
import React from "react";
const state = {
  giftCard: {
    giftCardDetails: {
      cardName: "Uber Gift Card",
      cardPoints: 12,
      cardCategory: "Ecommerce",
      cardRetailer: "Amazon",
      cardIssueDate: "2022-04-18T07:11:54.233Z",
      cardExpiryDate: "2025-01-01T00:00:00.000Z",
      cardCount: 91,
      cardImage: "https://source.unsplash.com/random/250*250?apple",
      cardVendor: "Uber",
      cardShortDesc: "10% OFF",
      cardLongDesc:
        " Lorem ipsum dolor sit amet consectetur adipisicing elit. Molestiae quisquam tenetur alias ducimus nostrum voluptate, nam suscipit ipsa dolorem mollitia error soluta, distinctio esse. Earum, quo! Cupiditate dignissimos iste ratione.",
      cardComments: [],
      id: 1650265917967,
    },
    emailLoader: false,
    emailError: "",
  },
  auth: {
    user: {},
    isAuthenticated: false,
  },
  loading: {
    loading: false,
  },
};
describe("Gift card details test suite", () => {
  const useSelectorMock = jest.spyOn(reactRedux, "useSelector");
  const useDispatchMock = jest.spyOn(reactRedux, "useDispatch");
  const useEffectMock = jest.spyOn(React, "useEffect");
  beforeEach(() => {
    useSelectorMock.mockReturnValue(state);
    useEffectMock.mockImplementation((f) => f());
    const loadSingleCard = jest.fn();
    useDispatchMock.mockReturnValue(loadSingleCard);
  });
  afterEach(() => {
    jest.restoreAllMocks();
  });
  const theme = createTheme();
  const middlewares = [thunk];
  const mockStore = configureStore(middlewares);

  const customRender = (location, type = "render", initialState = state) => {
    const history = createMemoryHistory();
    history.push(location);
    const store = mockStore(initialState);
    if (type === "render") {
      return render(
        <Provider store={store}>
          <ThemeProvider theme={theme}>
            <Router location={history.location} navigator={history}>
              <GiftCardDetails />
            </Router>
          </ThemeProvider>
        </Provider>
      );
    } else {
      return mount(
        <Provider store={store}>
          <ThemeProvider theme={theme}>
            <Router location={history.location} navigator={history}>
              <GiftCardDetails />
            </Router>
          </ThemeProvider>
        </Provider>
      );
    }
  };

  it("Should create a snapshot of the GiftCardDetails  page", () => {
    const { asFragment } = customRender("card/1");
    expect(asFragment()).toMatchSnapshot();
  });

  it("Should call the loadSingleCard function inside the dispatch method", () => {
    const loadSingleCard = jest.fn();
    const view = customRender("card/1", "mount", state);
    const store = mockStore(state);
    store.dispatch(loadSingleCard);
    expect(loadSingleCard).toHaveBeenCalled();
  });

  it("Should contain the gift card number that been load from the selectors state", () => {
    customRender("card/1");
    expect(
      screen.getByText(`#GIFT-${state.giftCard.giftCardDetails.id + 2022}`)
    ).toBeInTheDocument();
  });

  it("Should be have Show empty comment if we don't have any comment added", () => {
    customRender("card/1");
    expect(
      screen.getByText(/No comments yet. Be the first to comment./i)
    ).toBeInTheDocument();
  });

  it("Should be so the 'Add Comment' button if we have authenticated", () => {
    customRender("card/1", "render", {
      ...state,
      auth: { ...state.auth, isAuthenticated: true },
    });
    expect(
      screen.getByRole("button", {
        name: /Add Comment/i,
      })
    ).toBeInTheDocument();
  });

  it("Should call the handelSubmitted method when we click on the add comment", async () => {
    customRender("card/1", "render", {
      ...state,
      auth: { ...state.auth, isAuthenticated: true },
    });
    const addCommentButton = screen.getByRole("button", {
      name: /Add Comment/i,
    });
    expect(
      screen.getByRole("button", {
        name: /Add Comment/i,
      })
    ).toBeInTheDocument();
    const commentElement = screen.getByLabelText(/Your valuable comments.../i);
    fireEvent.change(commentElement, { target: { value: "test comment" } });
    fireEvent.click(addCommentButton);
    await waitFor(() => {
      expect(commentElement.value).toBe("test comment");
    });
  });
});
