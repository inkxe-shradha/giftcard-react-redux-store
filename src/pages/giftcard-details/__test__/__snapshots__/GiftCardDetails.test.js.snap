// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gift card details test suite Should create a snapshot of the GiftCardDetails  page 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthXl mt-5 css-19r6kue-MuiContainer-root"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 makeStyles-paper-1 css-1ps6pg7-MuiPaper-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 css-480o17-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 css-11l5t4l-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-12 css-1idn90j-MuiGrid-root"
            >
              <img
                alt="Card Img Details"
                class="img-fluid img-thumbnail"
                width="100%"
              />
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-12 css-1idn90j-MuiGrid-root"
            >
              <hr
                class="MuiDivider-root MuiDivider-fullWidth mb-2 css-9mgopn-MuiDivider-root"
              />
              <button
                class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root  css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                Send The Gift
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
              <table
                class="table table-dark mt-1 table-borderless"
              >
                <tbody>
                  <tr>
                    <td>
                      <span
                        class="MuiTypography-root MuiTypography-p css-1kbbegv-MuiTypography-root"
                      >
                        <strong>
                          Gift Card Number
                        </strong>
                      </span>
                    </td>
                    <td>
                      #GIFT-2123
                    </td>
                  </tr>
                  <tr>
                    <td>
                      <span
                        class="MuiTypography-root MuiTypography-p css-1kbbegv-MuiTypography-root"
                      >
                        <strong>
                          Gift Card Value
                        </strong>
                      </span>
                    </td>
                    <td>
                      $ 
                    </td>
                  </tr>
                  <tr>
                    <td>
                      <span
                        class="MuiTypography-root MuiTypography-p css-1kbbegv-MuiTypography-root"
                      >
                        <strong>
                          Gift Card Expiry Date
                        </strong>
                      </span>
                    </td>
                    <td>
                      April 29th 2022
                    </td>
                  </tr>
                  <tr>
                    <td>
                      <span
                        class="MuiTypography-root MuiTypography-p css-1kbbegv-MuiTypography-root"
                      >
                        <strong>
                          Available Cards
                        </strong>
                      </span>
                    </td>
                    <td>
                      <span
                        class="badge bg-gradient badge-pill bg-danger"
                      />
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-8 css-so71yv-MuiGrid-root"
        >
          <div
            class="MuiBox-root css-0"
          >
            <h4
              class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
            >
               - 
              <small>
                 
              </small>
            </h4>
          </div>
          <div
            class="mt-2 MuiBox-root css-0"
          >
            <span
              class="MuiTypography-root MuiTypography-body css-37jdci-MuiTypography-root"
            />
          </div>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth my-2 css-9mgopn-MuiDivider-root"
          />
          <div
            class="makeStyles-maxImageHeight-2 MuiBox-root css-0"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              Rating & Reviews
            </h6>
          </div>
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation10 css-10kr5gr-MuiPaper-root"
          >
            <div
              class="my-2"
            >
              <h6
                class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
              >
                Add Comment
              </h6>
              <span
                class="MuiTypography-root MuiTypography-body css-37jdci-MuiTypography-root"
              >
                Please add your comment here.
              </span>
            </div>
            <div
              class="dv-star-rating"
              style="display: inline-block; position: relative;"
            >
              <input
                class="dv-star-rating-input"
                id="rating_5"
                name="rating"
                style="display: none; position: absolute; margin-left: -9999px;"
                type="radio"
                value="5"
              />
              <label
                class="dv-star-rating-star dv-star-rating-empty-star"
                for="rating_5"
                style="float: right; cursor: pointer; color: rgba(255, 255, 255, 1);"
              >
                <i
                  style="font-style: normal;"
                >
                  ★
                </i>
              </label>
              <input
                class="dv-star-rating-input"
                id="rating_4"
                name="rating"
                style="display: none; position: absolute; margin-left: -9999px;"
                type="radio"
                value="4"
              />
              <label
                class="dv-star-rating-star dv-star-rating-empty-star"
                for="rating_4"
                style="float: right; cursor: pointer; color: rgba(255, 255, 255, 1);"
              >
                <i
                  style="font-style: normal;"
                >
                  ★
                </i>
              </label>
              <input
                class="dv-star-rating-input"
                id="rating_3"
                name="rating"
                style="display: none; position: absolute; margin-left: -9999px;"
                type="radio"
                value="3"
              />
              <label
                class="dv-star-rating-star dv-star-rating-empty-star"
                for="rating_3"
                style="float: right; cursor: pointer; color: rgba(255, 255, 255, 1);"
              >
                <i
                  style="font-style: normal;"
                >
                  ★
                </i>
              </label>
              <input
                class="dv-star-rating-input"
                id="rating_2"
                name="rating"
                style="display: none; position: absolute; margin-left: -9999px;"
                type="radio"
                value="2"
              />
              <label
                class="dv-star-rating-star dv-star-rating-empty-star"
                for="rating_2"
                style="float: right; cursor: pointer; color: rgba(255, 255, 255, 1);"
              >
                <i
                  style="font-style: normal;"
                >
                  ★
                </i>
              </label>
              <input
                class="dv-star-rating-input"
                id="rating_1"
                name="rating"
                style="display: none; position: absolute; margin-left: -9999px;"
                type="radio"
                value="1"
              />
              <label
                class="dv-star-rating-star dv-star-rating-empty-star"
                for="rating_1"
                style="float: right; cursor: pointer; color: rgba(255, 255, 255, 1);"
              >
                <i
                  style="font-style: normal;"
                >
                  ★
                </i>
              </label>
            </div>
            <div
              aria-label="your comments"
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="outlined-pass-comments-static"
                id="outlined-pass-comments-static-label"
              >
                Your valuable comments...
                <span
                  aria-hidden="true"
                  class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                >
                   *
                </span>
              </label>
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-8ewcdo-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <textarea
                  aria-invalid="false"
                  class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                  id="outlined-pass-comments-static"
                  name="comment"
                  required=""
                  rows="4"
                  style="height: 0px; overflow: hidden;"
                />
                <textarea
                  aria-hidden="true"
                  class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                  readonly=""
                  style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
                  tabindex="-1"
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-1ftyaf0"
                  >
                    <span>
                      Your valuable comments... *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <button
              class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root btn-comment mt-2 css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Add Comment
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
